Add limits on sampling factors

sohomdatta1 opened the pull request:
Invalid sampling factors can lead to a divide by zero error
which could eventually lead to a infinite loop.

Also fixed failing tests + add a new test for invalid sampling
factors.

fixes #105

patrickhulce commented:
> Also fixed failing tests

There shouldn't be failing tests. Can you elaborate? https://github.com/jpeg-js/jpeg-js/runs/3080117536?check_suite_focus=true

sohomdatta1 commented:
> > Also fixed failing tests
> 
> There shouldn't be failing tests. Can you elaborate? https://github.com/jpeg-js/jpeg-js/runs/3080117536?check_suite_focus=true
The max resolution and max memory tests kept hitting the sampling factor error (which I added in the current patch) before the respective errors that were being tested causing both of those tests to fail.

The changes wrt to the old tests just replace the old binary blobs with ones that do not hit the sampling factor error but rather only the specific errors being tested.

Wrt to the fix itself, I was basing it on how the libjpeg-turbo and libjpeg libraries handled this particular scenario:
https://github.com/libjpeg-turbo/libjpeg-turbo/blob/1f55ae7b0fa3acc348a630171617d0e56d922b68/jdinput.c#L64-L78

Based on their documentation, the sampling rate must be between 1 and 4 according to the JPEG standard (thus side-stepping the divide by zero issue).
https://github.com/libjpeg-turbo/libjpeg-turbo/blob/9abeff46d87bd201a952e276f3e4339556a403a3/libjpeg.txt#L1138-L1146

patrickhulce commented:
Awesome thank you for the very clear explanations @sohomdatta1! 🙏 

Would you mind adding the link to https://github.com/libjpeg-turbo/libjpeg-turbo/blob/9abeff46d87bd201a952e276f3e4339556a403a3/libjpeg.txt#L1138-L1146 along with our `= 1` changes and we'll merge this 🎉

sohomdatta1 commented:
> Awesome thank you for the very clear explanations @sohomdatta1! pray
> 
> Would you mind adding the link to https://github.com/libjpeg-turbo/libjpeg-turbo/blob/9abeff46d87bd201a952e276f3e4339556a403a3/libjpeg.txt#L1138-L1146 along with our `= 1` changes and we'll merge this tada

Made the change :)

patrickhulce reviewed (approved):
thank you so much :)
